packages/core/src/types.ts

```diff
@@ -10,34 +10,56 @@ export type WidgetState = {
 };
 export const WIDGET_DEFAULT_STATE: WidgetState = { showChat: false };
 
-// ## TrackParticipantPair Types
-export type TrackParticipantPair = {
-  track: TrackPublication;
+// ## TrackBundle Types
+
+export type TrackBundleSubscribed = {
   participant: Participant;
+  source: Track.Source;
+  publication: TrackPublication;
+  track: NonNullable<TrackPublication['track']>;
 };
-/**
- * A participant with no published tracks.
- * @remarks
- * Useful if you want to have a representation for participants without a published track.
- */
-export type TrackParticipantPlaceholder = {
-  track: undefined;
+
+export type TrackBundlePublished = {
+  participant: Participant;
   source: Track.Source;
+  publication: TrackPublication;
+};
+
+export type TrackBundlePlaceholder = {
   participant: Participant;
+  source: Track.Source;
 };
-export type MaybeTrackParticipantPair = TrackParticipantPair | TrackParticipantPlaceholder;
 
-// ### TrackParticipantPair Type Predicates
-export function isTrackParticipantPair(
-  item: MaybeTrackParticipantPair,
-): item is TrackParticipantPair {
-  return item.track !== undefined;
+export type TrackBundle = TrackBundleSubscribed | TrackBundlePublished;
+export type TrackBundleWithPlaceholder =
+  | TrackBundleSubscribed
+  | TrackBundlePublished
+  | TrackBundlePlaceholder;
+
+// ### TrackBundle Type Predicates
+export function isTrackBundle(bundle: TrackBundleWithPlaceholder): bundle is TrackBundle {
+  return (
+    isTrackBundleSubscribed(bundle as TrackBundle) || isTrackBundlePublished(bundle as TrackBundle)
+  );
+}
+
+export function isTrackBundleSubscribed(bundle: TrackBundle): bundle is TrackBundleSubscribed {
+  return bundle.hasOwnProperty('track');
 }
 
-export function isTrackParticipantPlaceholder(
-  item: MaybeTrackParticipantPair,
-): item is TrackParticipantPlaceholder {
-  return item.track === undefined && item.hasOwnProperty('source');
+export function isTrackBundlePublished(bundle: TrackBundle): bundle is TrackBundlePublished {
+  return bundle.hasOwnProperty('publication') && !bundle.hasOwnProperty('track');
```

Comment:
wondering if this is the right way around.
Just from naming I would expect every `BundleSubscribed` to also be `BundlePublished`

Reply:
Fair point. You would `isTrackBundle` for that. This is only to narrow down the type to the exact `TrackBundlePublished` type. But this is also not used outside so we can make it internal.